### PR TITLE
Document build_constraint in extensions.md

### PR DESCRIPTION
**Pull request by blegat**
Closes https://github.com/JuliaOpt/JuMP.jl/issues/1740

### CODE REVIEW

File: docs/src/extensions.md [hunk 1]
```diff
@@ -1,5 +1,93 @@
...
 TODO: How to extend JuMP: discussion on different ways to build on top of JuMP.
 How to extend JuMP's macros and how to avoid doing this.
+
+## Extending MOI
+
+TODO: Create new MOI function/sets, how to use it in JuMP, create new bridge
+
+## Extending JuMP macros
+
+In order to provide the most convenient syntax for the user to create variables,
```
Comment (mlubin) [inline docs/src/extensions.md]: "Most convenient" implies that we think that the macros are the best possible syntax. Maybe "a convenient syntax".

File: docs/src/extensions.md [hunk 2]
```diff
@@ -1,5 +1,93 @@
...
+`AbstractConstraint` with a corresponding [`add_constraint`](@ref) method (see
+[Adding `add_constraint` methods](@ref)).
+
+```@docs
+build_constraint
+```
+
+#### Shapes
+
+To convert the constraint to a `VectorConstraint`, it needs to be converted to
```
Comment (mlubin) [inline docs/src/extensions.md]: To convert which constraint? Maybe say instead, "shapes allow vector constraints, which are represented as flat vectors in MOI, to retain a matrix shape at the JuMP level".

File: docs/src/expressions_draft.md [hunk 3]
```diff
@@ -0,0 +1,12 @@
+### Defining a new MOI function
```
Comment (mlubin) [inline docs/src/expressions_draft.md]: Is this file referenced anywhere?

Comment (blegat) [reply docs/src/expressions_draft.md]: Oups, it shouldn't have been added

File: docs/src/extensions.md [hunk 4]
```diff
@@ -1,5 +1,91 @@
...
+
+## Extending the `@constraint` macro
+
+The [`@constraint`](@ref) macro always calls the same three functions:
+* `parse_constraint`: is called at parsing time, it parses the constraint
+  expression and returns a [`build_constraint`](@ref) call expression;
+* [`build_constraint`](@ref): given the functions and sets involved in the
+  constraints, it returns a `AbstractConstraint`;
+* [`add_constraint`](@ref): given the model, the `AbstractConstraint`
+  constructedi in [`build_constraint`](@ref) and the constraint name, it stores
```
Comment (mlubin) [inline docs/src/extensions.md]: typo: "constructedi"

**Pull request merged**